src/Style.hs

```diff
@@ -300,130 +314,145 @@ initSpec = StySpec { spType = C.PointT,
+genFns f stmts varmap = foldl (f varmap) [] stmts
 
+-- | 'procBlock' is called by 'getDictAndFns'. 'getDictAndFns' would fold this function on a list of blocks, a.k.a. a Style program, and accumulate objective/constraint functions, and a dictionary of geometries to be rendered.
+-- TODO: first split `stmts` into assigns, objs, and contrs and call procConstrFn and procObjFn with `map`?
+procBlock :: (Autofloat a) =>
+    StyContext a ->
+    Block ->
+    StyContext a
+-- Case 1: @global@ block - only objectives and constraints are allowed, and
+--- there is no need to compute any variable map
+procBlock (dict, objFns, constrFns) ((Selector C.AllT []) : [], stmts) =
+    let keys          = M.keys dict
+        identityMap   = M.fromList $ zip keys keys
+        newObjFns     = concatMap (procObjFn identityMap [])    stmts
+        newConstrFns  = concatMap (procConstrFn identityMap []) stmts
+    in (dict, objFns ++ newObjFns, constrFns ++ newConstrFns)
+
+-- Case 2: only one selector, no need to compute combinations.
+-- this function will process: (1) objectives, (2) constraints and (3)
+-- assignments
+procBlock (dict, objFns, constrFns) (selector : [], stmts) =
+    let mapsAndSpecs  = dict `matchWith` selector
+        newDict       = foldl (addShapes stmts) dict $ mapsAndSpecs
+        varmaps       = map fst $ mapsAndSpecs
+        newObjFns     = concatMap (genFns procObjFn stmts)    varmaps
+        newConstrFns  = concatMap (genFns procConstrFn stmts) varmaps
+    in (newDict, objFns ++ newObjFns, constrFns ++ newConstrFns)
+    where
         -- For the block, add each statement to the spec and insert the spec in the overall stydict
-        -- Only process assignment statements on matched specs, not the cartesion product of them
-        updateSpec :: StyDict -> (VarMap, StySpec) -> StyDict -- stmts is only external param
-        updateSpec dct (varmap, spec) =
-            let newSpec = foldl (procAssign varmap) spec stmts in -- procAssign is only called here
-            M.insert (spId newSpec) newSpec dct
-
-        -- Add each object's spec (that has been selected by the block's selectors) to the stydict
-        newDict :: StyDict
-        newDict = foldl updateSpec dict $ concat selectedSpecs
-
-        -- validMap :: t (k, [Char]) -> Bool
-        validMap ts = and . trs "result" . fst $ foldl
+        addShapes :: [Stmt] -> StyDict -> (VarMap, StySpec) -> StyDict
+        addShapes statements dct (varmap, spec) =
+            let newSpec = foldl (procAssign varmap) spec statements
+            in M.insert (spId newSpec) newSpec dct
+
+-- Case 3: multiple selectors, which in the current Style languages means
+-- we would take a subset of the cartesian product among the selectors.
+-- TEMP: the system does not support assignment statements in these blocks,
+-- because it is not clear that which Substance object the shapes associate with
+procBlock (dict, objFns, constrFns) (selectors, stmts) =
```

Comment:
design Q: do we want to consider other ways to join/combine selectors? e.g. `Set X | Point X { … }` (matches a set or a point) if so, we'd need to come up with use cases for that, and see how the Style compiler would need to generalize